### PR TITLE
[heartbeat] States and Improved Errors

### PR DESCRIPTION
**Pull request by andrewvc**
Fixes https://github.com/elastic/beats/issues/32163 , corresponding mapping changes for synthetics package in https://github.com/elastic/integrations/pull/4023

Adds a notion of state across checks, with flapping as a bonus. At a high level this PR does the following:

1. Adds new root level `state` fields
2. Enhances the `ecserr` package and types to make them more testable and usable
3. Refactors timeout, http status, and could not connect errors to use the new `ecserr` package to make testing this PR/feature easier (these are the easiest types of errors to replicate) with lightweight monitors.
4. Adds support for the standard `mage goIntegTest` task, already supported by CI that thus far has been a noop for heartbeat.
5. Adds a notion of flapping states, in addition to up / down states.
6. Automatically connects to ES to retrieve the last state value for the given monitor when a monitor first starts, this is necessary to continue the previous state across restarts of heartbeat
7. Replaces the `add_observer_metadata` processor with a new `heartbeat.location` global setting and `location` per monitor setting. This lets us set a location ID (which is then set to `observer.name`. See details below:

**Note: flapping is currently disabled**

Per the discussion in the review, it's a complex feature, let's add it in a follow-up

## What are states, and how are they implemented here?

The main goal of this PR is to resolve https://github.com/elastic/beats/issues/32163 , which this goes, but it also recognizes that the goal of grouping errors is a subset of the more general problem of grouping both up and down states. It's useful to group both since it's useful to see something like:

| State | Duration   | Reason     |
|-------|------------|------------|
| UP    | 18 hours   |            | 
| DOWN  | 30 minutes | status 400 |   |
| Up    | 1 month    |            |

Hence, the introduction of the various `state.*` fields, which group contiguous blocks of 'up' and 'down' states together. 

A sample of the `state.*` fields can be seen below:

```json5
{
  "state": { // new state field in addition to existing monitor fields
            // globally unique ID for this state, this ID is sortable as a timestamp 
            // to speed up aggregations. The format is id-timestampMsHex-serialHex
            // which is more compact than a UUID, and also chronologically sortable
            "id": "dummy-182a27ea210-2dc",
            // when this state first started, with this we can see when the first event in the
            // state occurred without having to retrieve that event
            "started_at": "2022-08-15T12:13:04.2721958-05:00",
            // number of milliseconds this state has been active for
            "duration_ms": 4655149,
            // number of checks that have occurred within this state
            // broken out by up/down. Flapping states will have non-zero values for both up/down
            "checks": 2290,
            "up": 0,
            "down": 2290
             // status of the state, which can be 'up', 'down', or 'flapping'
            // usually identical to monitor.status except in the case of a flapping monitor
            "status": "down",
            // the last FLAPPING_THRESHOLD-1 checks, used to reconstruct flapping state
            // when resuming state from ES
            "flap_history": [], 
            // The prior state, the nice thing about `state.ends` is that these states do not change
            // since they are complete, so they are easy to query / aggregate since the values are stable
            // in actual use these are only attached to events with `state.checks: 1` so they appear
            // exactly once
            "ends": {
              "started_at": "2022-08-15T12:12:57.1792082-05:00",
              "duration_ms": 5069,
              "status": "flap",
              "up": 3,
              "down": 1,
              "flap_history": null, // omitted on ends states since it's just dead-weight
              "id": "dummy-182a27e865b-2db",
              "checks": 4,
              "ends": null // we don't recurse end states
            },
          },
}
```

### Notes on location

The new location field can be set as follows:

```json5
#globally
heartbeat.location:
  id: "us-east-1a"
  geo:
    name: "US-East Coast"
    location: "44.123, 45.12345"
```

```json5
heartbeat.monitors:
- type: http
  id: my-monitor
  urls: "http://elastic.co"
  run_from:
    id: "us-east-1a"
      geo:
        name: "US-East Coast"
        location: "44.123, 45.12345"
```

### Notes on flapping states

The new flapping state serves an important purpose, to reduce the cardinality of states for unstable sites. This is important for UX and UI reasons, since having large numbers of states to visualize in a list is a key thing we'd like to improve.

The flapping threshold in this PR is hard coded to the number 7. This number is equivalent to the number of consecutive identical 'up' or 'down' states the algorithm uses to determine whether a monitor is stable or not. As an example, if a monitor experiences 7 consecutive up checks, followed by seven consecutive down checks it will be reflected as a single up state of 7 checks followed by a single down state of 7 checks. If, by contrast, there are 7 consecutive up checks followed by 6 consecutive down checks, then a single up check there will be two consecutive states, of `up` followed by `flapping`; if 6 consecutive down checks were to follow the last of these new events would constitute a new `down` state following the flapping state, since the monitor would now be stable.

Please see the unit tests for monitor states for additional more nuanced detail. I don't think it makes sense to expose this to users yet, though we could in the future. It's a bit complex to explain, and I think this is a good starting point. We'll likely want to tweak this algorithm in the future, but we that could be done in a follow-up. One concern I have is that it could take a while for monitors to recover if they run infrequently. 

It should be noted that flapping checks __start__ as a simple up or down check, but change into a flapping check if they see a different result before the flapping threshold is hit. So the most recent `state.status` is the only accurate value that should be used. It is also for this reason that two consecutive up or down states cannot happen, but multiple consecutive flapping states could happen if after what looks like a recover instability occurs again. We may want to tweak this to allow for shorter stable states. Again, I think this could happen in a flapping follow-up.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Create some monitors and verify that the `state` field is populated as expected in Elasticsearch. Since heartbeat just re-uses the ES connection, the easiest way is just to use the `elastic` user for your ES output, or just set `cloud.id` and `cloud.auth`. 

You could try the following kibana dashboard (exported and compressed as a zip file) to make an easier time of it.

[states-dashboard-export.zip](https://github.com/elastic/beats/files/9457080/states-dashboard-export.zip)

### CODE REVIEW

**Comment by rdner** [comment]
@andrewvc is this PR still relevant?

File: x-pack/heartbeat.run/integ test x-pack_heartbeat.run.xml
Comment (andrewvc) [inline x-pack/heartbeat.run/integ test x-pack_heartbeat.run.xml:1]: Useful for intellij users, this makes the integration test seamless

File: x-pack/heartbeat/monitors/browser/synthexec/execmultiplexer.go
Comment (andrewvc) [inline x-pack/heartbeat/monitors/browser/synthexec/execmultiplexer.go:17]: Sort of an unrelated change, but since we only ever use `*ExecMultiPlexer` that should really be the receiver. I hit a weird bug in some tests that this fixes.

File: x-pack/heartbeat/monitors/browser/synthexec/synthexec_test.go
Comment (andrewvc) [inline x-pack/heartbeat/monitors/browser/synthexec/synthexec_test.go:217]: more incorrectness that was due to be fixed, though it's a bit pedantic I shoehorned it in

**Comment by andrewvc** [comment]
I'm realizing that this is missing one piece of functionality, which is geolocation awareness, which is pretty important. We should segment everything by geo. I'll add that soon, but I think it's still reviewable in its current state, I'll try and make the geo stuff into a single commit to make it easy to review.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (dominiqueclarke) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate.go:46]: I'm not sure that "now" is the right started at in all cases. 

Consider the complexity introduced with flapping:

Let's say a monitor runs every 5 minutes, went down at 10:30 and came back up at 10:35. Because of the flap threshold, we'd wait another 2 checks, until 10:45, to consider this down state closed.

In actuality, the down event only occurred between 10:30 and 10:35. 

That means the next state actually begins at 10:35, not 10:45, when we finally determined that the monitor was indeed NOT flapping. This function is called from `transitionTo`, which is only called after the flapping threshold of checks is complete. Because of this, the new state would register as starting at 10:45, not 10:35, leaving a gap in states of 10 minutes.

Once we determine that the monitor is indeed not flapping, it seems we need to backdate the new state to start at the first stable check (ex: 10:35), and include first `FLAPPING_THRESHOLD` number of checks as part of the history of the new state.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (dominiqueclarke) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate.go:128]: It seems like all monitors will have a period of time where they are considered flapping in this logic? Will all monitors, once the state changes from up to down or down to up, be considered flapping until the `FLAPPING_THRESHOLD` number of checks with the new status?

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (dominiqueclarke) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate.go:151]: One we determine a monitor is not flapping, I'm not sure it will be enough just to assign ends to the last state. The state needs to be adjusted to include the actual duration of the event.

Consider my above example, where a monitor is down for 5 minutes, but took 15 minutes to ensure it was not flapping. If the `ends` data is not updated, the duration for the state will report as 15 minutes, rather than 5 minutes.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate.go:46]: That's a good point, and it's a limitation of this flapping implementation. For a short incident the flapping check does the wrong thing.

It might be best to disable flapping for now (we could just set a flap threshold of 1) and do a follow-up PR to address flapping. It's a complex topic, but I don't want to hold this up for that.

There's a choice in here in terms of whether we do want to backdate / rewrite history. The current code makes the choice to never do that. That could create a lot of confusing UX. I imagine in a UI it'd be confusing to see things shift in time etc. I'd prioritize stability of events over accuracy if that makes sense. I'm open to discussion here however.

Another option would be to require at least 3 state changes within a period of time to enter flapping. A quick outage and recovery should probably never result in a flap.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate.go:128]: It doesn't actually work that way (which can be seen in the real world data where we are more often than not in up or down, but not flap states), the condition above where `s.Status == newStatus` ensures that. In english I'd say that with `s.Checks < FlappingThreshold` we're in the eligible to flap zone.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate.go:151]: It's a good point to discuss, see my point earlier, I propose we disable flapping for now and discuss flapping in a follow-up PR. I'm apprehensive about an implementation where history can be rewritten, even the little bit we do now I don't like.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (dominiqueclarke) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate.go:46]: I think it's best to disable flapping for now and do a follow up PR, given it's complexity

> That could create a lot of confusing UX.

I think that there are UX challenges on both sides of this conversation. I think the complexity of flapping currently could also cause a confusing, and potentially inaccurate UX. If we can't determine if the state is "closed" until we've ruled out flapping, then the duration of the state always has the potential to be `flappingThreshold * monitorSchedule` greater than the actual duration of the error event. I think this would be even more confusing for users, to see down events that are longer than their systems were actually down.

With the flapping complexity, the `ends` optimization becomes less helpful. It would require querying against `ends`, then checking if the last `flappingThreshold` number of checks was consistent. If so, consider the event duration as `duration - (flappingThreshold * schedule)`. 

I want to avoid bloating the timings of events when displaying in the UI as much as possible, to give our users better fidelity of the actual duration and timing of error events. 

So I think it's best to table flapping for now, and continue thinking through the complexity at another time.

**Comment by andrewvc** [comment]
Additional changes (added to top desc as well):

1. Disabled flapping for now, tests are in place but a hard coded value determines whether it is enabled now, it is currently set to false
2. New settings for geolocation added at `heartbeat.location` and per-monitor on `location`.
3. We now scope location lookups to geolocation (and type for efficiency)

Will add more tests and docs for location soon, hopefully tomorrow

**Comment by andrewvc** [comment]
This is now ready for a full review, having incorporated initial feedback from @dominiqueclarke . 

It now has significantly enhanced functional-ish tests via many enhancements to the scenarios system / framework.

File: heartbeat/monitors/factory.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/factory.go:199]: This caught me off-guard, can we add comment about Fleet using monitor specific location whereas as a User I can use top level to avoid duplicating it over all places.

File: heartbeat/monitors/factory.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/factory.go:267]: Neat use of singleton ❤️

File: heartbeat/monitors/factory.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/factory.go:268]:
```suggestion
				logp.L().Warnf("could not add heartbeat geo info: %w", err)
```

File: heartbeat/monitors/wrappers/monitorstate/esloader.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/wrappers/monitorstate/esloader.go:55]: Shouldn't we look at record counts instead of doing by range? Reason being this varies depending on how frequent individual monitors are scheduled to run.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate.go:90]: shouldn't this be <=0 ?

**Review by vigneshshanmugam — COMMENTED** [review]
Checkpoint, Still reviewing the main monitor state details.

File: heartbeat/hbtestllext/isdefs.go
Comment (andrewvc) [reply heartbeat/hbtestllext/isdefs.go:84]: I was thinking this function was more strict, hence the exact quality

File: heartbeat/monitors/factory.go
Comment (andrewvc) [reply heartbeat/monitors/factory.go:199]: good point, will push that comment shortly

File: heartbeat/monitors/wrappers/monitorstate/esloader.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/esloader.go:55]: data streams / time series indexes really should always be paired with a time range if possible since it dramatically improves efficiency. ES can discard rotated indexes entirely if they don't contain that range. So, rather than having to examine say 365 indices with daily rotation it'd only query one.

Since 6h is well beyond the interval the UI supports, or a reasonable user would choose, I thought this made sense. Glad to up it to 12 or 24h but 6 seems quite large IMHO.

Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/esloader.go:55]: I think by record counts I think you meant by timestamp descending, which this query does do.

**Comment by andrewvc** [comment]
I believe I've addressed the incremental you reported so far @vigneshshanmugam thanks for starting on this review!

File: heartbeat/monitors/wrappers/monitorstate/monitorstate_test.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate_test.go:56]: nit: we could remove the 300ms upper bound here.I see no harm in that.

File: heartbeat/monitors/wrappers/monitorstate/tracker_test.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/wrappers/monitorstate/tracker_test.go:40]: nit: use recordStableSeries?

File: x-pack/heartbeat/scenarios/README.md
Comment (vigneshshanmugam) [inline x-pack/heartbeat/scenarios/README.md:8]: Love the flexibility of these new tests framework.

File: x-pack/heartbeat/scenarios/framework/framework.go
Comment (vigneshshanmugam) [inline x-pack/heartbeat/scenarios/framework/framework.go:270]: Do we need this explicitly set here? I thought we handle this on our loader interface? fakeloader should have this.

**Review by vigneshshanmugam — COMMENTED** [review]
Looks good overall, I havent digged much in to the flapping states and thrershold when we switch them as we are disabling that particular feature for this PR.

File: heartbeat/monitors/wrappers/monitorstate/esloader.go
Comment (vigneshshanmugam) [reply heartbeat/monitors/wrappers/monitorstate/esloader.go:55]: Thanks for the details Andrew. It makes sense, I was going for like
- Don't care about the range
- Just get the entries that we are interested in. 

But I did not know about the efficiency improvement we get with the time ranges in ES. ++ 6 hours is way too large.

File: x-pack/heartbeat/scenarios/framework/framework.go
Comment (andrewvc) [reply x-pack/heartbeat/scenarios/framework/framework.go:270]: yup, will remove, leftover from an old version

File: heartbeat/monitors/wrappers/monitorstate/tracker.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/tracker.go:79]: I think this would be fine for `run_once` mode because it should be constrained to `1^2+2^2+500*3=6500ms` total delay + the time to actually exec the requests

Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/tracker.go:79]: To add a little more detail here, the internet can be a little flakey sometimes, so I think it's a good thing to do a sort of minimal retry along the lines we do here for `run_once` as well. Also, the regular case is similar to `run_once` in that we don't want to impede the startup of heartbeat if ES is down, but it's OK to delay it... a little bit. So, it would make sense to use the same logic for both.

File: heartbeat/monitors/wrappers/monitorstate/monitorstate_test.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate_test.go:56]: My worry is that there could be a bug in how we calculate the duration in the future. Let me up it to say 900 though, that's very forgiving.

File: heartbeat/monitors/wrappers/monitorstate/tracker_test.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/tracker_test.go:40]: It's in another package. It's quite irritating that you can't use stuff from other package's `_test.go` files without extra complexity. Given that this is just here, I think it's simplest to just have the loop here :(

Sometimes I just add the test helpers to a `testhelpers.go` file in that package to expose them, but it always feels a bit hacky.

File: heartbeat/monitors/wrappers/monitorstate/esloader.go
Comment (andrewvc) [reply heartbeat/monitors/wrappers/monitorstate/esloader.go:55]: Yeah, unfortunately ES has no way, without a timestamp, to just get the latest AFAIK. My understanding is that if you want to get the latest it gets the most recent document from every index, then finds the most recent one. With a time range it excludes all but the most recent index

**Comment by andrewvc** [comment]
@vigneshshanmugam FYI 6ecac3e fixes an interesting little bug, where heartbeat would try to connect to ES with the default options when non-ES outputs were defined. This was caught by the python tests

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (vigneshshanmugam) [inline heartbeat/monitors/wrappers/monitorstate/monitorstate.go:47]: Need to account for missing `RunFrom` config. This is what I see in states

```
  "state": {
    "started_at": "2022-09-13T11:22:16.489384-07:00",
    "up": 0,
    "ends": null,
    "id": "my-monitor-%!s(*config.LocationWithID=<nil>)-183381669a9-0",
    "duration_ms": 20003,
    "status": "down",
    "checks": 3,
    "down": 3,
    "flap_history": []
  },
```

We can default to `unknown` location ?

File: x-pack/heartbeat/scenarios/framework/fakeloader.go
Comment (vigneshshanmugam) [inline x-pack/heartbeat/scenarios/framework/fakeloader.go:30]: Move this logic to monitorstate.go?

File: heartbeat/monitors/wrappers/monitorstate/monitorstate.go
Comment (vigneshshanmugam) [reply heartbeat/monitors/wrappers/monitorstate/monitorstate.go:47]: Also it should be RunFrom.id, since its using interface now.

File: heartbeat/config/config.go
Comment (vigneshshanmugam) [inline heartbeat/config/config.go:42]: Should we also change this to `RunFrom`

**Review by vigneshshanmugam — APPROVED** [review]
LGTM with pending comments, This is exciting 🎉 

Tested with lightweight and browser jobs. State seems to be populated correctly.

File: heartbeat/config/config.go
Comment (andrewvc) [reply heartbeat/config/config.go:42]: Yeah, I was thinking this was a good top level name, but now I think you're right, consistency is more important.

File: heartbeat/monitors/factory.go
Comment (emilioalvap) [reply heartbeat/monitors/factory.go:269]: @andrewc a bit of a late question, sorry about that. Isn't this statement assuming that there'll be only one location across all available monitors? In that case, would it be a fair assumption to make?

**Pull request merged**
